src/systems/hydrodynamics/Hydrodynamics.cc

```diff
@@ -363,40 +331,50 @@ void Hydrodynamics::PreUpdate(
-  Cmat(3, 5) =   this->dataPtr->paramMdotQ * state(4);
-  Cmat(4, 0) =   this->dataPtr->paramZdotW * state(2);
-  Cmat(4, 2) = - this->dataPtr->paramXdotU * state(0);
-  Cmat(4, 3) =   this->dataPtr->paramNdotR * state(5);
-  Cmat(4, 5) = - this->dataPtr->paramKdotP * state(3);
-  Cmat(5, 0) =   this->dataPtr->paramZdotW * state(2);
-  Cmat(5, 1) =   this->dataPtr->paramXdotU * state(0);
-  Cmat(5, 3) = - this->dataPtr->paramMdotQ * state(4);
-  Cmat(5, 4) =   this->dataPtr->paramKdotP * state(3);
+  // for the under water vehicle's additional DOF. We are just considering
+  // diagonal terms here. Have yet to add the cross terms here. Also note, since
+  // $M_a(0,0) = \dot X_u $ , $M_a(1,1) = \dot Y_v $ and so forth, we simply
+  // load the stability deravitives from $M_a$.
+  Cmat(0, 4) = - this->dataPtr->Ma(2, 2) * state(2);
+  Cmat(0, 5) = - this->dataPtr->Ma(1, 1) * state(1);
+  Cmat(1, 3) =   this->dataPtr->Ma(2, 2) * state(2);
+  Cmat(1, 5) = - this->dataPtr->Ma(0, 0) * state(0);
+  Cmat(2, 3) = - this->dataPtr->Ma(1, 1) * state(1);
+  Cmat(2, 4) =   this->dataPtr->Ma(0, 0) * state(0);
+  Cmat(3, 1) = - this->dataPtr->Ma(2, 2) * state(2);
+  Cmat(3, 2) =   this->dataPtr->Ma(1, 1) * state(1);
+  Cmat(3, 4) = - this->dataPtr->Ma(5, 5) * state(5);
+  Cmat(3, 5) =   this->dataPtr->Ma(4, 4) * state(4);
+  Cmat(4, 0) =   this->dataPtr->Ma(2, 2) * state(2);
+  Cmat(4, 2) = - this->dataPtr->Ma(0, 0) * state(0);
+  Cmat(4, 3) =   this->dataPtr->Ma(5, 5) * state(5);
+  Cmat(4, 5) = - this->dataPtr->Ma(3, 3) * state(3);
+  Cmat(5, 0) =   this->dataPtr->Ma(2, 2) * state(2);
+  Cmat(5, 1) =   this->dataPtr->Ma(0, 0) * state(0);
+  Cmat(5, 3) = - this->dataPtr->Ma(4, 4) * state(4);
+  Cmat(5, 4) =   this->dataPtr->Ma(3, 3) * state(3);
   const Eigen::VectorXd kCmatVec = - Cmat * state;
 
-  // Damping forces (Fossen P. 43)
-  Dmat(1, 1)
-    = - this->dataPtr->paramYv - this->dataPtr->paramYvv * abs(state(1));
-  Dmat(0, 0)
-    = - this->dataPtr->paramXu - this->dataPtr->paramXuu * abs(state(0));
-  Dmat(2, 2)
-    = - this->dataPtr->paramZw - this->dataPtr->paramZww * abs(state(2));
-  Dmat(3, 3)
-    = - this->dataPtr->paramKp - this->dataPtr->paramKpp * abs(state(3));
-  Dmat(4, 4)
-    = - this->dataPtr->paramMq - this->dataPtr->paramMqq * abs(state(4));
-  Dmat(5, 5)
-    = - this->dataPtr->paramNr - this->dataPtr->paramNrr * abs(state(5));
+  // Damping forces
+  for(int i = 0; i < 6; i++)
+  {
+    for(int j = 0; j < 6; j++)
+    {
+      auto coeff = - this->dataPtr->stabilityLinearTerms[i * 6 + j];
+      for(int k = 0; k < 6; k++)
+      {
+        auto index = i * 36 + j * 6 + k;
+        auto absCoeff =
+          this->dataPtr->stabilityQuadraticAbsDerivative[index] * abs(state(k));
+        coeff -= absCoeff;
+        auto velCoeff =
+          this->dataPtr->stabilityQuadraticDerivative[index] * state(k);
+        coeff -= velCoeff;
+      }
+
+      Dmat(i, j) = coeff;
+    }
+  }
```

Comment:
@arjo129 meta: I'll start by saying I know next to nothing about underwater vessel modelling. That said, I can't relate this equations to those in Fossen's books (ie. Guidance of Ocean Vehicles and Marine Control Systems). I see you are including all cross terms, but none of the models proposed includes a pure quadratic term nor the means to estimate most of the rest (which begs the question, why bother?). 

Mind to explain?

Reply:
[So the LRAUV's original model does](https://github.com/osrf/lrauv/wiki/Vehicle-specifications-and-hydrodynamic-parameters), we had not implemented it as we were interested in getting the hydrodynamics plugin up and running early on. It is often hard to estimate and verify the cross terms, however there are definitely techniques that can be used.